sqlmesh/core/state_sync/engine_adapter.py

```diff
@@ -1149,29 +1226,62 @@ def _get_snapshot_intervals(
             rows = self._fetchall(query.where(where))
             interval_ids.update(row[0] for row in rows)
 
+            # Pending restatement intervals are aggregated per (name, version) as opposed to snapshot ID
+            pending_restatement_intervals: t.Dict[t.Tuple[str, str], Intervals] = defaultdict(list)
+            last_seen_identifier_per_version: t.Dict[t.Tuple[str, str], str] = {}
+
             intervals: t.Dict[t.Tuple[str, str, str], Intervals] = defaultdict(list)
             dev_intervals: t.Dict[t.Tuple[str, str, str], Intervals] = defaultdict(list)
             for row in rows:
-                _, name, identifier, version, start, end, is_dev, is_removed = row
+                (
+                    _,
+                    name,
+                    identifier,
+                    version,
+                    start,
+                    end,
+                    is_dev,
+                    is_removed,
+                    is_pending_restatement,
+                ) = row
                 intervals_key = (name, identifier, version)
                 target_intervals = intervals if not is_dev else dev_intervals
                 if is_removed:
                     target_intervals[intervals_key] = remove_interval(
                         target_intervals[intervals_key], start, end
                     )
+                elif is_pending_restatement:
+                    pending_restatement_intervals[(name, version)] = merge_intervals(
+                        [*pending_restatement_intervals[(name, version)], (start, end)]
+                    )
+                    last_seen_identifier_per_version[(name, version)] = identifier
                 else:
                     target_intervals[intervals_key] = merge_intervals(
                         [*target_intervals[intervals_key], (start, end)]
                     )
+                    if not is_dev:
+                        # Remove all pending restatement intervals recorded before the current interval has been added
+                        pending_restatement_intervals[(name, version)] = remove_interval(
+                            pending_restatement_intervals[(name, version)], start, end
+                        )
+                        last_seen_identifier_per_version[(name, version)] = identifier
 
             for name, identifier, version in {**intervals, **dev_intervals}:
+                # Only set the pending restatement intervals for the last seen snapshot per version
+                # FIXME: Remove this logic once intervals are stored per (name, version)
+                pending_restatement_intervals_for_snapshot = (
```

Comment:
are there any issues with only storing this on the latest one?

Reply:
Haven't found any yet. This is because we add these to each snapshot that share the same version. So the identifier doesn't really matter